### DISCUSSION TITLE
Root component for the BaseGame?

### DISCUSSION THREAD

**Discussion by st-pasha** [flame-engine/flame#899] [Ideas]
I've been looking at flame's code, and noticed that there are lots of similarities between `BaseGame` and `BaseComponent`: they both have a `ComponentSet`, facilities to add, remove and rebalance children.

So I was wondering, what if, instead, `BaseGame` had a single "root" `BaseComponent` that would serve as the container for all other components? This could have the following potential benefits:
  - avoid code repetition between `BaseGame` and `BaseComponent`;
  - the components in the game scene now form a proper tree, making tree-traversal algorithms easier;
  - there won't be a feature mismatch between what's possible with the `BaseGame` and the `BaseComponent`. For example, currently one of them has `propagateToChildren` while the other `changePriorities`, however, both methods should be applicable to any component tree;
  - when `BaseGame` has a single root, certain features are easier to implement:
     - `onGameResize` can be handled by removing the current root and replacing it with a new root recreated from scratch to fit the new size;
     - game navigation (similar to Flutter's `Navigator`) can be accomplished by replacing the current root with a new root that implements the UI for a new page. For example, consider a hero travelling on the world map and deciding to open his/her inventory. If the inventory is full-screen, I could just replace the root `WorldComponent` with the `InventoryComponent`, and then switch them back when the inventory is closed. I can even make the WorldComponent run at 1/10th of the normal clock speed while the inventory is open by running `world.update(dt/10)` in my inventory's update function.

I don't see any downsides with this approach (but obviously my knowledge of the engine is very shallow), and it won't be a breaking change. A new method `BaseComponent replaceRoot(BaseComponent)` will be added.

What do you guys think?

**Comment by spydon** [comment]
I think it is a very good idea, we actually discussed it last week in the team. Another feature that will be the result of this is that you can add a game to a game.

**Reply by spydon** [reply]
Started the work for this in #906 :)
